[Merged by Bors] - feat(Data/Polynomial/Module) : define an R[X] module, given a linear map.

rmhi opened the pull request:
Given an element `a` in an `R`-algebra `A` and an `A`-module `M`,
define an `R[X]`-module `Module.AEval R M a`, whose elements correspond to
elements of `M`, and where the action of `f : R[X]` is `f • m = aeval a f • m`. 
Equivalently, `X • m = a • m`.

This module is abbreviated to `Module.AEval' φ` in the special case that `A` is the algebra of `R`-linear maps and `φ : M → M` is an `R`-linear map.

This is needed in #7419.

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -350,3 +352,79 @@ theorem comp_smul (p p' : R[X]) (q : PolynomialModule R M) :
+Then `Module.AEval R M a` is the `R[X]`-module with carrier `M`,
+where the action of `f : R[X]` is `f • m = (aeval a f) • m`.
+-/
+structure AEval (R M: Type*) {A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
```
I don't think there is any need to create a new structure here, you can just use a type synonym. In practice you define `AEval R M a := M` (as a type) and then you provide the `R`-module structure by hand. You can have a look of what we do for [PiLp](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/PiLp.html#PiLp), do define the product with the Lp metric.

rmhi replied:
Thank you for the really quick response.

In my original PR on this (7419) I did define the module as a type synonym, and was advised that it would be clearer to make it a single-field structure. I will change it back.

riccardobrasca replied:
Well, I am not 100% sure. @eric-wieser what do you think?

alexjbest commented:
This sounds like it has the same goal as https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteType.html#modulePolynomialOfEndo, this PR should redefine that material in terms of this I think. (I think using a type synonym is probably a nicer way to go, and it looks like you added more api here)

riccardobrasca commented:
> This sounds like it has the same goal as https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteType.html#modulePolynomialOfEndo, this PR should redefine that material in terms of this I think. (I think using a type synonym is probably a nicer way to go, and it looks like you added more api here)

Why is this in `Mathlib.RingTheory.FiniteType`?! Nice found BTW.

alexjbest commented:
I think because its only used to prove the result below! But it really shouldn't be indeed

alexjbest commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -350,3 +351,62 @@ theorem comp_smul (p p' : R[X]) (q : PolynomialModule R M) :
+
+namespace AEval
+
+instance : AddCommMonoid <| AEval R M a     := by assumption
+instance : Module R <| AEval R M a          := by assumption
+instance : Module A <| AEval R M a          := by assumption
+instance : IsScalarTower R A <| AEval R M a := by assumption
```
It took me a minute to work out how these proofs even work, I think using `inferInstanceAs (AddCommMonoid M)` is probably cleaner

alexjbest commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -350,3 +351,62 @@ theorem comp_smul (p p' : R[X]) (q : PolynomialModule R M) :
+where the action of `f : R[X]` is `f • m = (aeval a f) • m`.
+-/
+@[nolint unusedArguments]
+def AEval (R M: Type*) {A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
```
```suggestion
def AEval (R M : Type*) {A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
```

rmhi commented:
> This sounds like it has the same goal as https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteType.html#modulePolynomialOfEndo, this PR should redefine that material in terms of this I think. (I think using a type synonym is probably a nicer way to go, and it looks like you added more api here)

Thanks very much for that; I didn't think it was already in Mathlib. I'll try doing what you suggest.

rmhi commented:
> I think it should still go in Polynomial/Module if that works though, just add an import to that from the finitetype file, as Riccardo says FiniteType is a non-obvious place for this construction to live

Thanks, I've switched that around now. I also needed to swap the imports, so that now FiniteType imports Polynomial/Module, whereas previously it was the other way around. I've moved my code to the start of Polynomial/Module, because it is now used later down that file. I have left Vasconcelos' theorem in FiniteType, but its proof now uses `AEval'` instead of `ModulePolynomialOfEndo`. This avoids the creation of instances at the start of the proof, and also a `have` statement, which is a lemma for `AEval'`.

There is one thing which I wonder about. On line 729 of FiniteType, I need to quote the incidence `AEval'.Finite_of_Finite`. I had hoped that lean would find that incidence for itself, and I'm a little concerned that it can't.

alexjbest commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -179,16 +150,21 @@ theorem induction_linear {P : PolynomialModule R M → Prop} (f : PolynomialModu
 #align polynomial_module.polynomial_module PolynomialModule.polynomialModule
 
+lemma smul_def (f : R[X]) (m : PolynomialModule R M) :
+    f • m = (aeval  (Finsupp.lmapDomain M R Nat.succ)) f m:= by
```
```suggestion
    f • m = aeval (Finsupp.lmapDomain M R Nat.succ) f m:= by
```

alexjbest commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -14,17 +14,58 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
 
 This is defined as a type alias `PolynomialModule R M := ℕ →₀ M`, since there might be different
 module structures on `ℕ →₀ M` of interest. See the docstring of `PolynomialModule` for details.
```
Can you add a couple of lines here in the module header about `AEval` also?

rmhi replied:
Thanks for pointing that out.

alexjbest commented:
Ok looks good to me now, I left two last small comments, but after those I think this should be good to go

alexjbest commented:
maintainer merge

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,64 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+Then `Module.AEval R M a` is the `R[X]`-module with carrier `M`,
+where the action of `f : R[X]` is `f • m = (aeval a f) • m`.
 -/
+@[nolint unusedArguments]
+def AEval (R M : Type*) {A : Type*} [CommSemiring R] [Semiring A] [Algebra R A]
+    [AddCommMonoid M] [Module A M] [Module R M] [IsScalarTower R A M] (_ : A) := M
```
As we discussed in #7419, I think this should be a one-field structure; semireducible type aliases are messy in Lean 4, so we should only use them where the cost is worth it.

If you do really want to keep this as not a structure, then can you add the canonical `LienarEquiv` between `M` and `AEval R M a`?

rmhi replied:
It's seems that you are asking me to undo a change requested by another reviewer, and his request was that I undo an earlier change requested you you. I don't have the experience to make a judgement on which way is better.

riccardobrasca replied:
OK, it is definitely my fault. Can you please go back to the approach with a structure? Sorry for the waste of time.

eric-wieser replied:
I'm happy with the compromise of sticking with the synonym and just adding the isomorphism

rmhi replied:
Thanks for the compromise. I'll sort this out some time in the next few days (I'll be teaching all day today).

rmhi replied:
is this ok now?

eric-wieser replied:
I pushed some cleanup

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,73 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+instance : Module R <| AEval R M a                      := inferInstanceAs (Module R M)
+instance : Module A <| AEval R M a                      := inferInstanceAs (Module A M)
+instance : IsScalarTower R A <| AEval R M a             := inferInstanceAs (IsScalarTower R A M)
+instance [AddCommGroup M] : AddCommGroup <| AEval R M a := inferInstanceAs (AddCommGroup M)
```
This instance is bad, it assumes `[AddCommMonoid M]` and `[AddCommGroup M]` simultaneously.

rmhi replied:
Thank you for finding that mistake.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,73 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
 
-open Polynomial BigOperators
+instance : AddCommMonoid <| AEval R M a                 := inferInstanceAs (AddCommMonoid M)
+instance : Module R <| AEval R M a                      := inferInstanceAs (Module R M)
+instance : Module A <| AEval R M a                      := inferInstanceAs (Module A M)
```
Please give these and the other module instance explicit names, to make it easy to tell them apart in the infoview. Perhaps `instModuleOrig`, `instModuleEval`, and `instModulePolynomial`?

rmhi replied:
Thank you.

eric-wieser reviewed (commented):
Sorry for coming in so late with this observation, but there's an instance diamond I'm worried about:
```lean
/-- In the special case where `A = R[X]` (eg multivariate polynomials), we have two conflicting
actions which agree only in the special case of of `a = X`. -/
example (a : R[X]) (h : a ≠ X):
    (instModuleEval a : Module R[X] (AEval R R[X] a)) ≠
    (instModulePolynomial a  : Module R[X] (AEval R R[X] a)) := by
  intro H
  apply h
  replace H := congr_fun₂ (congr_arg (@HSMul.hSMul _ _ _ <| @instHSMul _ _ ·.toSMul) H)
  dsimp [instModuleEval, instModulePolynomial, smul_def] at H
  replace H := H X (1 : R[X])
  simpa using H.symm
```
I think either we:
* Put a warning in the docstring that it is not safe to use `AEval` in this way (and commit the `example` that demonstrates it.
* Remove the `Module A (AEval R M a)` instance, especially if you weren't actually using it

rmhi commented:
> Sorry for coming in so late with this observation, but there's an instance diamond I'm worried about:
> 
> ```lean
> /-- In the special case where `A = M = R[X]` (eg multivariate polynomials), we have two conflicting
> actions which agree only in the special case of of `a = X`. -/
> example (a : R[X]) (h : a ≠ X):
>     (instModuleEval a : Module R[X] (AEval R R[X] a)) ≠
>     (instModulePolynomial a  : Module R[X] (AEval R R[X] a)) := by
>   intro H
>   apply h
>   replace H := congr_fun₂ (congr_arg (@HSMul.hSMul _ _ _ <| @instHSMul _ _ ·.toSMul) H)
>   dsimp [instModuleEval, instModulePolynomial, smul_def] at H
>   replace H := H X (1 : R[X])
>   simpa using H.symm
> ```
> 
> I think either we:
> 
> * Put a warning in the docstring that it is not safe to use `AEval` in this way (and commit the `example` that demonstrates it.
> * Remove the `Module A (AEval R M a)` instance, especially if you weren't actually using it

Many thanks for pointing this out. I have use this instance quite a lot (for example, in order to state smul_def, which is used in several places). I can see that this would cause problems, so I'll have a think about it.

eric-wieser commented:
> Many thanks for pointing this out. I have use this instance quite a lot (for example, in order to state smul_def, which is used in several places). I can see that this would cause problems, so I'll have a think about it.

One resolution here would be to state `smul_def` as `f • m = of (aeval a f • of.symm m)`.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,82 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+lemma X_smul (m : M) : (X : R[X]) • (of R M a m) = a • m := by simp [smul_def]
+
+instance instIsScalarTowerOrigPolynomial : IsScalarTower R R[X] <| AEval R M a where
+  smul_assoc r f m := by
```
I think you'll find this useful:
```suggestion
  smul_assoc r f m := of.symm.injective by
```
then you can probably lose the `change` below.

rmhi replied:
many thanks; I did find the change annoying.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,82 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+lemma smul_def (f : R[X]) (m : M) : f • (of R M a m) = aeval a f • m := by
+  rfl
+
+lemma X_smul (m : M) : (X : R[X]) • (of R M a m) = a • m := by simp [smul_def]
```
This one is also poorly-typed.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,96 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+
+lemma smul_def₁ (f : R[X]) (m : M) : f • (of R M a m) = of R M a (aeval a f • m) := by rfl
 
+lemma smul_def₂ (f : R[X]) (m : AEval R M a) :
```
I think this is a better name (and also that this is a good `simp` lemma:
```suggestion
@[simp] lemma of_symm_smul (f : R[X]) (m : AEval R M a) :
```

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,96 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+
+variable {R M}
+
+lemma smul_def₁ (f : R[X]) (m : M) : f • (of R M a m) = of R M a (aeval a f • m) := by rfl
```
This can be given a nicer name if you write it in reverse:
```suggestion
lemma of_aeval_smul (f : R[X]) (m : M) : of R M a (aeval a f • m) = f • of R M := rfl
```

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,96 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+@[simp] lemma of_symm_smul (f : R[X]) (m : AEval R M a) :
+    (of R M a).symm (f • m) = (aeval a f • (of R M a).symm m) := by rfl
 
+lemma X_smul₁ (m : M) : (X : R[X]) • (of R M a m) = of R M a (a • m) := by
+  rw [smul_def₁, aeval_X]
+
+lemma X_smul₂ (m : AEval R M a) : (of R M a).symm ((X : R[X]) • m) = a • (of R M a).symm m := by
+  rw [smul_def₂, aeval_X]
```
Can you rename these in the same way to match the above?

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,102 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+
+variable {R M}
+
+lemma of_aeval_smul (f : R[X]) (m : M) : of R M a (aeval a f • m) = f • of R M a m:= rfl
```
```suggestion
lemma of_aeval_smul (f : R[X]) (m : M) : of R M a (aeval a f • m) = f • of R M a m := rfl
```

eric-wieser replied:
This one also still applies. I'd recommend not resolving suggestions in future until you see them marked as [outdated], otherwise you might have forgotten to apply them!

rmhi replied:
sorry about that. I think I must have messed up with git somehow.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,102 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+lemma of_aeval_smul (f : R[X]) (m : M) : of R M a (aeval a f • m) = f • of R M a m:= rfl
 
+@[simp] lemma of_symm_smul (f : R[X]) (m : AEval R M a) :
+    (of R M a).symm (f • m) = (aeval a f • (of R M a).symm m) := by rfl
```
```suggestion
    (of R M a).symm (f • m) = aeval a f • (of R M a).symm m := rfl
```

rmhi replied:
Thank you.

eric-wieser replied:
This comment still applies; `rfl` is better than `by rfl` because it trains `dsimp`.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,102 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+lemma AEval'_def : AEval' φ = AEval R M φ := rfl
+lemma AEval'.X_smul_of (m : M) : (X : R[X]) • AEval'.of φ m = AEval'.of φ (φ m) :=
+  AEval.X_smul_of _ _
+lemma AEval'.of_symm_X_smul (m : AEval' φ) : (AEval'.of φ).symm ((X : R[X]) • m) = φ m :=
```
This one is not well-typed, it should be 
```suggestion
lemma AEval'.of_symm_X_smul (m : AEval' φ) : (AEval'.of φ).symm ((X : R[X]) • m) = φ ((AEval'.of φ).symm m) :=
```

rmhi replied:
yes, I am certainly seeing the advantages of the one-field structure. I had considered making this change, but I have been using the lemma from right to left (in a place where I should really have been more rigorous). I'll resolve the issue.

eric-wieser commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,102 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+`Module.AEval'` is defined as a special case of `Module.AEval` in which the `R`-algebra is
+`M →ₗ[R] M`. Lemmas involving `Module.AEval` may be applied to `Module.AEval'`.
+-/
+abbrev AEval' := AEval R M φ
```
```suggestion
abbrev AEval' : Type _ := AEval R M φ
```

eric-wieser reviewed (approved):
This is looking good now, thanks!

rmhi commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,102 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+/-
+`Module.AEval'` is defined as a special case of `Module.AEval` in which the `R`-algebra is
+`M →ₗ[R] M`. Lemmas involving `Module.AEval` may be applied to `Module.AEval'`.
+-/
```
thank you

rmhi commented:
Many thanks for all you help!

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,101 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+namespace Module
+/--
+Suppose `a` is an element of an `R`-algebra `A` and `M` is an `A`-module.
+Loosely speaking, `Module.AEval R M a` is the `R[X]`-module with element `m : M`,
```
```suggestion
Loosely speaking, `Module.AEval R M a` is the `R[X]`-module with elements `m : M`,
```

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,101 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+    AddCommGroup <| AEval R M a := inferInstanceAs (AddCommGroup M)
 
-open Polynomial BigOperators
+variable {R A M} [CommSemiring R] [Semiring A] (a : A) [Algebra R A] [AddCommMonoid M] [Module A M]
```
It seems the `instance` just above this line uses exactly these variables. Can you move the `variable` line before the `instance` to make the declaration more readable?

rmhi replied:
This isn't quite true. The instance on line 41 assumes [AddCommGroup M], not [AddCommMonoid M]. If I moved the variables higher up then that instance would get an extra hypothesis.

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,101 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+
+lemma of_aeval_smul (f : R[X]) (m : M) : of R M a (aeval a f • m) = f • of R M a m:= rfl
 
+@[simp] lemma of_symm_smul (f : R[X]) (m : AEval R M a) :
```
Are you sure we want this to be a `simp` lemma? The RHS even contains `.symm`...

eric-wieser replied:
I'd say yes; if you have `(of R M a).symm` on the outside, you probably want to push it in to end up with an equation in `M` rather than `AEval R M a`. This is consistent with things like `Multiplicative.ofAdd_smul`.

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,101 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+where `φ : M →ₗ[R] M`.
+-/
+abbrev AEval'.of : M ≃ₗ[R] AEval' φ := AEval.of R M φ
+lemma AEval'_def : AEval' φ = AEval R M φ := rfl
+lemma AEval'.X_smul_of (m : M) : (X : R[X]) • AEval'.of φ m = AEval'.of φ (φ m) :=
+  AEval.X_smul_of _ _
+lemma AEval'.of_symm_X_smul (m : AEval' φ) :
+    (AEval'.of φ).symm ((X : R[X]) • m) = φ ((AEval'.of φ).symm m) := AEval.of_symm_X_smul _ _
+instance [Finite R M] : Finite R[X] <| AEval' φ := inferInstance
```
```suggestion
lemma AEval'_def : AEval' φ = AEval R M φ := rfl
lemma AEval'.X_smul_of (m : M) : (X : R[X]) • AEval'.of φ m = AEval'.of φ (φ m) :=
  AEval.X_smul_of _ _
  
lemma AEval'.of_symm_X_smul (m : AEval' φ) :
    (AEval'.of φ).symm ((X : R[X]) • m) = φ ((AEval'.of φ).symm m) := AEval.of_symm_X_smul _ _

instance [Finite R M] : Finite R[X] <| AEval' φ := inferInstance
```

riccardobrasca commented on Mathlib/Data/Polynomial/Module.lean:
```diff
@@ -15,16 +15,101 @@ In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R
+
+namespace AEval
+
+instance instAddCommMonoid : AddCommMonoid <| AEval R M a := inferInstanceAs (AddCommMonoid M)
+instance instModuleOrig : Module R <| AEval R M a := inferInstanceAs (Module R M)
+instance instFiniteOrig [Finite R M] : Finite R <| AEval R M a := inferInstanceAs (Finite R M)
+instance instModulePolynomial : Module R[X] <| AEval R M a := compHom M (aeval a).toRingHom
```
```suggestion
instance instAddCommMonoid : AddCommMonoid <| AEval R M a := inferInstanceAs (AddCommMonoid M)

instance instModuleOrig : Module R <| AEval R M a := inferInstanceAs (Module R M)

instance instFiniteOrig [Finite R M] : Finite R <| AEval R M a := inferInstanceAs (Finite R M)

instance instModulePolynomial : Module R[X] <| AEval R M a := compHom M (aeval a).toRingHom
```

riccardobrasca reviewed (approved):
Sorry again for the long review time, but this is very good now, thanks! (I've left a couple of trivial comments.)

bors d+

rmhi commented:
bors r+
